server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/GitMirror.java

```diff
@@ -333,4 +377,327 @@ private Git openGit(File workDir) throws Exception {
+            case JSON:
+                final JsonNode oldJsonNode = oldContent != null ? Jackson.readTree(oldContent) : null;
+                final JsonNode newJsonNode = (JsonNode) entry.content();
+
+                // Upsert only when the contents are really different.
+                if (!Objects.equals(newJsonNode, oldJsonNode)) {
+                    // Use InsertText to store the content in pretty format
+                    final String newContent = newJsonNode.toPrettyString() + '\n';
+                    applyPathEdit(dirCache, new InsertText(pathString, inserter, newContent));
+                    return newContent.length();
+                }
+                break;
+            case TEXT:
+                final String sanitizedOldText = oldContent != null ?
+                                                sanitizeText(new String(oldContent, UTF_8)) : null;
+                final String sanitizedNewText = entry.contentAsText(); // Already sanitized when committing.
+                // Upsert only when the contents are really different.
+                if (!sanitizedNewText.equals(sanitizedOldText)) {
+                    applyPathEdit(dirCache, new InsertText(pathString, inserter, sanitizedNewText));
+                    return sanitizedNewText.length();
+                }
+                break;
+        }
+        return 0;
+    }
+
+    private static void applyPathEdit(DirCache dirCache, PathEdit edit) {
+        final DirCacheEditor e = dirCache.editor();
+        e.add(edit);
+        e.finish();
+    }
+
+    private static byte[] currentEntryContent(ObjectReader reader, TreeWalk treeWalk) throws IOException {
+        final ObjectId objectId = treeWalk.getObjectId(0);
+        return reader.open(objectId).getBytes();
+    }
+
+    private static void maybeEnterSubtree(
+            TreeWalk treeWalk, String remotePath, String path) throws IOException {
+        // Enter if the directory is under the remote path.
+        // e.g.
+        // path == /foo/bar
+        // remotePath == /foo/
+        if (path.startsWith(remotePath)) {
+            treeWalk.enterSubtree();
+            return;
+        }
+
+        // Enter if the directory is equal to the remote path.
+        // e.g.
+        // path == /foo
+        // remotePath == /foo/
+        final int pathLen = path.length() + 1; // Include the trailing '/'.
+        if (pathLen == remotePath.length() && remotePath.startsWith(path)) {
+            treeWalk.enterSubtree();
+            return;
+        }
+
+        // Enter if the directory is the parent of the remote path.
+        // e.g.
+        // path == /foo
+        // remotePath == /foo/bar/
+        if (pathLen < remotePath.length() && remotePath.startsWith(path + '/')) {
+            treeWalk.enterSubtree();
+        }
+    }
+
+    /**
+     * Removes {@code \r} and appends {@code \n} on the last line if it does not end with {@code \n}.
+     */
+    private static String sanitizeText(String text) {
```

Comment:
Q) Is it possible to use a common method for this logic? (with storage side)

Reply:
It seems like this comment wasn't addressed. Is it possible to dedup this method?

https://github.com/line/centraldogma/blob/0cf990c73622145dcbb148da9464211e50a6fd38/server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepository.java#L1166

Reply:
Ah forgot to leave a comment about this. 😄 
I didn't intentionally dedup logic because of this PR https://github.com/line/centraldogma/pull/681
Let me cleanup the logic in the PR! 😉